File: api/src/main/java/com/FinancialAI/service/TransactionService.java
package com.FinancialAI.service;

import com.FinancialAI.controller.request.TransactionRequest;
import com.FinancialAI.controller.request.TransactionUpdateRequest;
import com.FinancialAI.controller.response.TransactionResponse;
import com.FinancialAI.domain.Category;
import com.FinancialAI.domain.Transaction;
import com.FinancialAI.domain.User;
import com.FinancialAI.domain.enums.TransactionType;
import com.FinancialAI.mapper.TransactionMapper;
import com.FinancialAI.repository.CategoryRepository;
import com.FinancialAI.repository.TransactionRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
public class TransactionService {

    private final TransactionRepository transactionRepository;
    private final TransactionMapper transactionMapper;
    private final AuthService authService;
    private final CategoryRepository categoryRepository;

    public TransactionResponse includeTransaction(TransactionRequest transactionRequest) {
        User userLogged = authService.getUser();

        Category category = categoryRepository
                .findByIdAndUserIdOrGlobal(transactionRequest.categoryId(), userLogged.getId())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND,"Categoria não encontrada"));

        Transaction transaction = transactionMapper.toEntity(transactionRequest);

        transaction.setUser(userLogged);
        transaction.setCategory(category);

        transaction.setCreatedAt(LocalDateTime.now());
        transaction.setUpdatedAt(LocalDateTime.now());

        Transaction saved = transactionRepository.save(transaction);

        return transactionMapper.toResponse(saved);
    }

    public TransactionResponse updateTransaction(Long transactionId, TransactionUpdateRequest request) {
        User userLogged = authService.getUser();

        Transaction transaction = transactionRepository
                .findByIdAndUserId(transactionId, userLogged.getId())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND,"Transação não encontrada"));

        if (request.amount() != null) {
            transaction.setAmount(request.amount());
        }

        if (request.transactionDate() != null) {
            transaction.setTransactionDate(request.transactionDate());
        }

        if (request.description() != null) {
            transaction.setDescription(request.description());
        }

        if (request.transactionType() != null) {
            transaction.setTransactionType(request.transactionType());
        }

        if (request.categoryId() != null) {

            Category category = categoryRepository
                    .findByIdAndUserIdOrGlobal(request.categoryId(), userLogged.getId())
                    .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND,"Categoria não encontrada"));

            transaction.setCategory(category);
        }

        transaction.setUpdatedAt(LocalDateTime.now());

        Transaction saved = transactionRepository.save(transaction);

        return transactionMapper.toResponse(saved);
    }

    public void deleteTransaction(Long transactionId) {
        User userLogged = authService.getUser();

        Transaction transaction = transactionRepository
                .findByIdAndUserId(transactionId, userLogged.getId())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND,"Transação não encontrada"));

        transactionRepository.delete(transaction);
    }

    public Page<TransactionResponse> listTransactions(Long categoryId, TransactionType type, Integer month, Integer year, Pageable pageable) {
        User userLogged = authService.getUser();

        Page<Transaction> transactions = transactionRepository.findWithFilters(
                userLogged.getId(), categoryId, type, month, year, pageable
        );

        return transactions.map(transactionMapper::toResponse);
    }
}

File: app/financial-ai/src/app/services/transaction.service.ts
import { inject, Injectable, signal } from '@angular/core';
import { TransactionRequest, TransactionResponse, TransactionUpdateRequest } from '../models/transaction.interface';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable, tap } from 'rxjs';
import { PageResponse } from '../models/page.interface';

@Injectable({
  providedIn: 'root',
})
export class TransactionService {
  
  transactions = signal<PageResponse<TransactionResponse> | null>(null);
  public apiUrl = signal<String>('http://localhost:8080')

  private http = inject(HttpClient)

  public includeTransaction(request: TransactionRequest): Observable<TransactionResponse> {
    return this.http.post<TransactionResponse>(`${this.apiUrl()}/transactions`, request)
  }

  public updateTransaction(transactionId: number, request: TransactionUpdateRequest): Observable<TransactionResponse> {
    return this.http.put<TransactionResponse>(`${this.apiUrl()}/transactions/${transactionId}`, request)
  }

  public deleteTransaction(transactionId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl()}/transactions/${transactionId}`)
  }

  public listTransactions(
    page: number = 0, 
    size: number = 10, 
    sort: string = 'transactionDate,desc', 
    filters: any = {}
  ): Observable<PageResponse<TransactionResponse>> {

    let params = new HttpParams()
      .set('page', page)
      .set('size', size)
      .set('sort', sort);

    if (filters.categoryId) params = params.set('categoryId', filters.categoryId);
    if (filters.type) params = params.set('type', filters.type);
    if (filters.month) params = params.set('month', filters.month);
    if (filters.year) params = params.set('year', filters.year);

    return this.http.get<PageResponse<TransactionResponse>>(`${this.apiUrl()}/transactions`, { params })
      .pipe(
        tap((pageData) => this.transactions.set(pageData))
      );
  }
}
